**Exercise 1 (Lab 11)**

**Jayant Kumar Jha(23M1730)**


# **Q.a**

The problem can be mathematically expressed as,

\begin{equation*}
\text{Maximize } \sum_{i=1}^{n} v_i \cdot x_i
\end{equation*}

\begin{align*}
\text{subject to } \sum_{i=1}^{n} t_i \cdot x_i &\leq T \\
x_i &\in \{0, 1\} \quad \text{for } i = 1, 2, \ldots, n
\end{align*}

Here, $v_i$ is the value of the $i^{th} $ task and $t_i$ is the time associated to $i^{th}$ task. $T$ is the total time window.

$x_i$ is the variable for doing the task $i$. X takes value 0 if task is not done and 1 if task is done.


In [ ]:
# install pyomo library to collab enivronment
!pip -q install pyomo
# import pyomo and numpy library to collab session
from pyomo.environ import *
import numpy as np

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.7/12.7 MB 39.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.6/49.6 kB 3.6 MB/s eta 0:00:00


In [ ]:
# number of variables N(=9) and number of constraints M(=14)
N, M = 20, 1
c = np.array([10,25,33,53,15,70,52,45,32,30,20,5,17,50,37,18,71,35,60,90])
b = np.array([40*60])
IP_Model = ConcreteModel()

#coefficient of constraints
A = np.array([
    [180, 180, 150, 150, 130, 210, 160, 175, 90, 120, 110,60, 145, 180, 120,120, 300, 300, 240, 500]])

In [ ]:
IP_Model.x = Var(np.arange(N), domain = Binary,bounds=(0,1))
IP_Model.const = ConstraintList()
IP_Model.constraints = ConstraintList()
[IP_Model.constraints.add(sum(A[i][j]*IP_Model.x[j] for j in range(N)) <= b[i]) for i in range(M)]
IP_Model.objective = Objective(expr = sum(c[j]*IP_Model.x [j] for j in range(N)), sense = maximize)
IP_Model.pprint()

1 Var Declarations
    x : Size=20, Index={0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19}
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :     1 : False :  True : Binary
          1 :     0 :  None :     1 : False :  True : Binary
          2 :     0 :  None :     1 : False :  True : Binary
          3 :     0 :  None :     1 : False :  True : Binary
          4 :     0 :  None :     1 : False :  True : Binary
          5 :     0 :  None :     1 : False :  True : Binary
          6 :     0 :  None :     1 : False :  True : Binary
          7 :     0 :  None :     1 : False :  True : Binary
          8 :     0 :  None :     1 : False :  True : Binary
          9 :     0 :  None :     1 : False :  True : Binary
         10 :     0 :  None :     1 : False :  True : Binary
         11 :     0 :  None :     1 : False :  True : Binary
         12 :     0 :  None :     1 : False :  True : Binary
         13 :     0 :  None :    

In [ ]:
# installing cbc solver to collab session
!apt-get -q install -y -qq coinor-cbc

# optCBC is will be used to solver above pyomo framework using

optCBC = SolverFactory('cbc')
resultCBC = optCBC.solve(IP_Model)


# snip to code for a function to print optimal solution as an output
def OptResults(Model):
  # display solution
  print('\nObjective = ', Model.Objective())

  print('\nDecision Variables')
  for i in np.arange(N):
    print('x[',i,'] = ', Model.x[i].value)
  print('\nConstraints')
  Model.constraints.display()

Selecting previously unselected package coinor-libcoinutils3v5:amd64.
(Reading database ... 121753 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.11.4+repack1-2_amd64.deb ...
Unpacking coinor-libcoinutils3v5:amd64 (2.11.4+repack1-2) ...
Selecting previously unselected package coinor-libosi1v5:amd64.
Preparing to unpack .../1-coinor-libosi1v5_0.108.6+repack1-2_amd64.deb ...
Unpacking coinor-libosi1v5:amd64 (0.108.6+repack1-2) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.17.5+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.17.5+repack1-1) ...
Selecting previously unselected package coinor-libcgl1:amd64.
Preparing to unpack .../3-coinor-libcgl1_0.60.3+repack1-3_amd64.deb ...
Unpacking coinor-libcgl1:amd64 (0.60.3+repack1-3) ...
Selecting previously unselected package coinor-libcbc3:amd64.
Preparing to unpack .../4-coinor-libcbc3_2.10.7+ds1-1_amd64.deb ...
Unpacking coinor-libcbc3:

In [ ]:
ILP_ModelSolution = OptResults(IP_Model)


Objective =  <pyomo.core.base.block._generic_component_decorator object at 0x7f5a1669b970>

Decision Variables
x[ 0 ] =  0.0
x[ 1 ] =  0.0
x[ 2 ] =  1.0
x[ 3 ] =  1.0
x[ 4 ] =  0.0
x[ 5 ] =  1.0
x[ 6 ] =  1.0
x[ 7 ] =  1.0
x[ 8 ] =  1.0
x[ 9 ] =  1.0
x[ 10 ] =  0.0
x[ 11 ] =  0.0
x[ 12 ] =  0.0
x[ 13 ] =  1.0
x[ 14 ] =  1.0
x[ 15 ] =  0.0
x[ 16 ] =  1.0
x[ 17 ] =  0.0
x[ 18 ] =  1.0
x[ 19 ] =  1.0

Constraints
constraints : Size=1
    Key : Lower : Body   : Upper
      1 :  None : 2395.0 : 2400.0


In [ ]:
print(f'{IP_Model.objective()} is the maximum value of the tasks')

623.0 is the maximum value of the tasks


So the preffered tasks are Task ( 3,  4, 6,  7,  8,  9,  10,  14,  15,  17,  19,  20

# **Q.b.**

We are provided with N tasks, each having a specific value and duration. Additionally, a maximum time window (T) is set. The goal is to select tasks in a way that the total value is maximized within the given time frame.

The strategy involves examining all possible task combinations, computing their cumulative time and value. Only those combinations whose total duration does not exceed T are considered. Among these, the combination with the highest total value is selected.

**Optimal Substructure:** For each task, there are two possibilities to consider in creating an optimal selection:

**Case 1:** The task is included in the optimal selection.
**Case 2:** The task is excluded from the optimal selection.

**Solution Steps:**

**To maximize the value obtained from N tasks, we compare two scenarios:**

**Case 1** (including the Nth task): This is the value of the Nth task combined with the maximum value achievable from the remaining N-1 tasks within the reduced time window (T - duration of the Nth task).

**Case 2** (excluding the Nth task): This is the maximum value achievable from the N-1 tasks within the time window T.
If the duration of the Nth task exceeds T, it cannot be included, making Case 2 the only viable option.

In [ ]:
def recursion(T, time, val, n):

    # Base Case
    if n == 0 or T == 0:
        return 0

    # If Time of the nth task is
    # more than Time window T,
    # then this task cannot be included
    # in the optimal solution
    if (time[n-1] > T):
        return recursion(T, time, val, n-1)

    # return the maximum of two cases:
    # (1) nth task included
    # (2) not included
    else:
        return max(
            val[n-1] + recursion(
                T-time[n-1], time, val, n-1),
            recursion(T, time, val, n-1))

# end of function recursion



value = [10,25,33,53,15,70,52,45,32,30,20,5,17,50,37,18,71,35,60,90]
times = [180, 180, 150, 150, 130, 210, 160, 175, 90, 120, 110,60, 145, 180, 120,120, 300, 300, 240, 500]
T = 40*60
n = len(value)
print(recursion(T, times, value, n))


623


In [ ]:
def is_subset_sum_recursive(arr, n, target, subset=[]):
    # Base Cases
    if target == 0:
        print(subset)
        return True
    if n == 0 and target != 0:
        return False
    # If last element is greater than target, ignore it
    if arr[n - 1] > target:
        return is_subset_sum_recursive(arr, n - 1, target, subset)

    return (is_subset_sum_recursive(arr, n - 1, target, subset) or
            is_subset_sum_recursive(arr, n - 1, target - arr[n - 1], subset + [arr[n - 1]]))

In [ ]:
array = [10,25,33,53,15,70,52,45,32,30,20,5,17,50,37,18,71,35,60,90]
target = 623
print("for set having the value of tasks {10,25,33,53,15,70,52,45,32,30,20,5,17,50,37,18,71,35,60,90}\n")
result = is_subset_sum_recursive(array, len(array), target)
print(result)

for set having the value of tasks {10,25,33,53,15,70,52,45,32,30,20,5,17,50,37,18,71,35,60,90}

[60, 71, 18, 37, 50, 17, 20, 30, 32, 45, 52, 70, 53, 33, 25, 10]
True


In [ ]:
value_subset = [60, 71, 18, 37, 50, 17, 20, 30, 32, 45, 52, 70, 53, 33, 25, 10]
print("The task selected are:")
for i in range(len(value_subset)):
  if value_subset[i] in array:
    print("Task", array.index(value_subset[i])+1)

The task selected are:
Task 19
Task 17
Task 16
Task 15
Task 14
Task 13
Task 11
Task 10
Task 9
Task 8
Task 7
Task 6
Task 4
Task 3
Task 2
Task 1


# **Q.c**

In [ ]:
def dynamicProgramming(time, value, T, n):

    # base conditions
    if n == 0 or T == 0:
        return 0
    if t[n][T] != -1:
        return t[n][T]

    # choice diagram code
    if time[n-1] <= T:
        t[n][T] = max(
            value[n-1] + dynamicProgramming(
                time, value, T-time[n-1], n-1),
            dynamicProgramming(time, value, T, n-1))
        return t[n][T]
    elif time[n-1] > T:
        t[n][T] = dynamicProgramming(time, value, T, n-1)
        return t[n][T]

# Driver code

value = [10,25,33,53,15,70,52,45,32,30,20,5,17,50,37,18,71,35,60,90]
time = [180, 180, 150, 150, 130, 210, 160, 175, 90, 120, 110,60, 145, 180, 120,120, 300, 300, 240, 500]
T = 40*60
n = len(value )

# We initialize the matrix with -1 at first.
t = [[-1 for i in range(T + 1)] for j in range(n + 1)]
print('The max value is:')
print(dynamicProgramming(time, value , T, n))

The max value is:
623


**Initial Conditions:** The algorithm initiates with two primary conditions. If the count of items (n) is reduced to 0 or if the knapsack's maximum capacity (T) is depleted, the highest achievable value defaults to 0.

**Caching Strategy:** To enhance efficiency, the algorithm employs a caching technique, recording outcomes of smaller problems to prevent repeat computations. It verifies whether a solution has been previously calculated for the given set of items and knapsack capacity. If a solution exists, it retrieves the saved result without additional computation.

**Decision Process:**

The algorithm evaluates each item through a binary choice: either to incorporate the item into the knapsack or to leave it out.
Should the item's weight not exceed the knapsack's available capacity (T), inclusion is an option. In such instances, the algorithm recursively processes the remaining items and available capacity post-inclusion, adding the item's value to the resultant value from the recursion.
Conversely, if the item's weight surpasses the knapsack's residual capacity, inclusion is not feasible. The algorithm then proceeds to the subsequent item, excluding the current one.

**Outcome:** Ultimately, the algorithm determines the maximum achievable value by exhaustively exploring all item and capacity combinations, leveraging cached results for efficiency, and delivers the optimal value attainable within the constraints of knapsack capacity.

In summary, the algorithm systematically navigates through all item and capacity permutations, utilizes cached data to minimize repeat computations, and conclusively identifies the highest value attainable.

In [ ]:
array = [10,25,33,53,15,70,52,45,32,30,20,5,17,50,37,18,71,35,60,90]
target = 623
print("for set having the value of tasks {10,25,33,53,15,70,52,45,32,30,20,5,17,50,37,18,71,35,60,90}\n")
result = is_subset_sum_recursive(array, len(array), target)
# print(result)
value_subset = [60, 71, 18, 37, 50, 17, 20, 30, 32, 45, 52, 70, 53, 33, 25, 10]
print("The task selected are:")
for i in range(len(value_subset)):
  if value_subset[i] in array:
    print("Task", array.index(value_subset[i])+1)


for set having the value of tasks {10,25,33,53,15,70,52,45,32,30,20,5,17,50,37,18,71,35,60,90}

[60, 71, 18, 37, 50, 17, 20, 30, 32, 45, 52, 70, 53, 33, 25, 10]
The task selected are:
Task 19
Task 17
Task 16
Task 15
Task 14
Task 13
Task 11
Task 10
Task 9
Task 8
Task 7
Task 6
Task 4
Task 3
Task 2
Task 1


# **Q2.**

In [ ]:
# Task data: (task number, task value, task time)
tasks = [
    (1, 10, 180), (2, 25, 180), (3, 33, 150), (4, 53, 150),
    (5, 15, 130), (6, 70, 210), (7, 52, 160), (8, 45, 175),
    (9, 32, 90), (10, 30, 120), (11, 20, 110), (12, 5, 60),
    (13, 17, 145), (14, 50, 180), (15, 37, 120), (16, 18, 120),
    (17, 71, 300), (18, 35, 300), (19, 60, 240), (20, 90, 500)
]

# Sort tasks by their task value in descending order
sorted_tasks = sorted(tasks, key=lambda x: x[1], reverse=True)

total_time = 0
total_value = 0
selected_tasks = []

# Greedy selection of tasks
for task in sorted_tasks:
    if total_time + task[2] <= 2400:
        total_time += task[2]
        total_value += task[1]
        selected_tasks.append(task[0])

# Output total task value and sequence of tasks
print("Total Task Value:", total_value)
print("Sequence of Tasks the manager takes:", selected_tasks)


Total Task Value: 568
Sequence of Tasks the manager takes: [20, 17, 6, 19, 4, 7, 14, 8, 15, 18, 12]


Here we can observe that the optimal solution does not remain same. In this case it is 568 but previously it was 623.

# **Q3.**

In [ ]:
tasks = [
    (1, 10, 180, 10/180), (2, 25, 180, 25/180), (3, 33, 150, 33/150), (4, 53, 150, 53/150),
    (5, 15, 130, 15/130), (6, 70, 210, 70/210), (7, 52, 160, 52/160), (8, 45, 175, 45/175),
    (9, 32, 90, 32/90), (10, 30, 120, 30/120), (11, 20, 110, 20/110), (12, 5, 60, 5/60),
    (13, 17, 145, 17/145), (14, 50, 180, 50/180), (15, 37, 120, 37/120), (16, 18, 120, 18/120),
    (17, 71, 300, 71/300), (18, 35, 300, 35/300), (19, 60, 240, 60/240), (20, 90, 500, 90/500)
] #value by time ratio is also added in this

In [ ]:
# Sort tasks by their task value in descending order
sorted_tasks = sorted(tasks, key=lambda x: x[3], reverse=True)

total_time = 0
total_value = 0
selected_tasks = []

# Greedy selection of tasks
for task in sorted_tasks:
    if total_time + task[2] <= 2400:
        total_time += task[2]
        total_value += task[1]
        selected_tasks.append(task[0])

# Output total task value and sequence of tasks
print("Total Task Value:", total_value)
print("Sequence of Tasks the manager takes:", selected_tasks)

Total Task Value: 601
Sequence of Tasks the manager takes: [9, 4, 6, 7, 15, 14, 8, 10, 19, 17, 3, 11, 16, 2, 12]


Here we can observe that the optimal solution does not remain same. In this case it is 601 but previously it was 623 in $1^{st}$ question.

# **Q4.**

# **a)**

$K := \{ x \in \{0, 1\}^n : \sum_{i=1}^{n} a_i x_i \leq b \}
$

$
K_C := \{ x \in \{0, 1\}^n : \sum_{i \in C} x_i \leq |C| - 1, \text{ for every } C \}
$


The statement asserts that each viable solution within $K$ can also be considered viable for $K_C$. Demonstrating this involves proving that for any viable solution $x$ belonging to $K$, we can identify a minimal cover $C$ where $x$ also fits within $K_C$.
Here, $K$ denotes the set of solutions satisfying the condition that the total weighted sum of $a_i x_i$ is under a threshold $b$. Conversely, $K_C$ signifies the set of solutions where, for every minimal cover $C$, the quantity of chosen items doesn't surpass the size of $C$.
To assert that $K$ is a subset of $K_C$, we must ensure that for each viable solution in $K$, a minimal cover $C$ can be found where the sum of chosen items is less than the count of $C$ by one. However, establishing this relationship is challenging because not every viable solution in $K$ may correspond to a minimal cover $C$ with $|C| - 1$ greater than the sum of chosen items.
Thus, it cannot be conclusively stated that $K$ is a subset of $K_C$.
    

# **b)**
"$\textbf{$K_C \subseteq K$}$"

The statement proposes that each viable solution within $K_C$ can equally serve as a viable solution for $K$. To establish this claim, we must demonstrate that every viable solution $x$ found in $K_C$ aligns with a total weighted sum of $a_i x_i$ that remains below the threshold $b$.

$K_C$ is defined as the set of solutions where, given any minimal cover $C$, the count of chosen items falls short of $|C| - 1$. Meanwhile, $K$ encompasses solutions that ensure the total weighted sum of $a_i x_i$ does not surpass $b$.
The proposition that $K_C$ is included in $K$ is valid because any solution qualifying under $K_C$ inherently meets the criteria set by $K$. Specifically, if the count of items selected for every minimal cover $C$ is below $|C| - 1$, it implies that the aggregate weighted sum of these selected items will not breach the limit $b$.

Hence, the relationship $K_C \subseteq K$ is confirmed to be accurate.